components/eam/src/physics/crm/sam/crm_variance_transport.F90

```diff
@@ -0,0 +1,368 @@
+      do k = 1,nzm
+         do j = 1,ny
+            do i = 1,nx
+               do icrm = 1,ncrms
+                  tmp_t(icrm,i,j,k) = t(icrm,i,j,k) 
+                  tmp_q(icrm,i,j,k) = qv(icrm,i,j,k) + qcl(icrm,i,j,k) + qci(icrm,i,j,k)
+                  tmp_t(icrm,i,j,k) = tmp_t(icrm,i,j,k) - t_mean(icrm,k)
+                  tmp_q(icrm,i,j,k) = tmp_q(icrm,i,j,k) - q_mean(icrm,k)
+               end do
+            end do
+         end do
+      end do
+
+      call VT_filter( ncrms, tmp_t, t_vt_pert )
+      call VT_filter( ncrms, tmp_q, q_vt_pert )
+
+   else ! use total variance
+
+      !$acc parallel loop collapse(4) async(asyncid)
+      do k = 1,nzm
+         do j = 1,ny
+            do i = 1,nx
+               do icrm = 1,ncrms
+                  t_vt_pert(icrm,i,j,k) = t(icrm,i,j,k) - t_mean(icrm,k)
+                  tmp = qv(icrm,i,j,k) + qcl(icrm,i,j,k) + qci(icrm,i,j,k)
+                  q_vt_pert(icrm,i,j,k) = tmp           - q_mean(icrm,k)
+               end do
+            end do
+         end do
+      end do
+
+   end if
+
+   !----------------------------------------------------------------------------
+   ! calculate variance
+   !----------------------------------------------------------------------------
+   !$acc parallel loop collapse(4) async(asyncid)
+   do k = 1,nzm
+      do j = 1,ny
+         do i = 1,nx
+            do icrm = 1,ncrms
+               t_vt(icrm,k) = t_vt(icrm,k) + t_vt_pert(icrm,i,j,k) * t_vt_pert(icrm,i,j,k)
+               q_vt(icrm,k) = q_vt(icrm,k) + q_vt_pert(icrm,i,j,k) * q_vt_pert(icrm,i,j,k)
+            end do
+         end do
+      end do
+   end do
+
+   !$acc parallel loop collapse(2) async(asyncid)
+   do k = 1,nzm
+      do icrm = 1,ncrms
+         t_vt(icrm,k) = t_vt(icrm,k) * factor_xy
+         q_vt(icrm,k) = q_vt(icrm,k) * factor_xy
+      end do
+   end do
+
+   !----------------------------------------------------------------------------
+   ! clean up
+   !----------------------------------------------------------------------------
+   deallocate( t_mean )
+   deallocate( q_mean )
+
+   deallocate( tmp_t )
+   deallocate( tmp_q )
+
+end subroutine VT_diagnose
+
+!===============================================================================
+!===============================================================================
+subroutine VT_forcing(ncrms)
+   !----------------------------------------------------------------------------
+   ! Purpose: Calculate forcing for variance injection and apply limiters
+   !----------------------------------------------------------------------------
+   use crmdims,      only: crm_dx
+   use vars,         only: t
+   use microphysics, only: micro_field, index_water_vapor
+   implicit none
+
+   ! interface arguments
+   integer, intent(in) :: ncrms
+
+   ! local variables
+   real(crm_rknd), parameter   :: pert_scale_min = 1.0 - 0.1
+   real(crm_rknd), parameter   :: pert_scale_max = 1.0 + 0.1
```

Comment:
Comments should be used to explain what these parameters are and how they are set.  Perhaps it would be better to not hardwire these here.

Reply:
I agree in general about the hardwiring... but in this case the value doesn't seem to make much difference, so I really doubt we'd want to be adjusting this via the namelist. I suppose it might come in handy in some odd situation where we are having trouble getting past a crash for a production run. I'm just gonna leave them in place for now (with comments) and we can turn them into namelist variables pretty quickly if there's a need.

Reply:
That sounds good to me, adding comments is sufficient.